src/main/java/org/currencygoldexchangeapp/handlers/ExchangeRateFileReaderHandler.java

```diff
@@ -0,0 +1,161 @@
+            }
+
+            return exchangeRatesFromFile;
+        }
+    }
+
+    public List<String> getErrorMessages() {
+        return new ArrayList<>(errorMessages.values());
+    }
+
+    private boolean processExchangeRateLine(String line, int lineNumber) {
+        String[] parts = line.split(" ");
+
+        if (parts.length == 4) {
+            String sourceCurrency = parts[0].trim();
+            String amountAsString = parts[1].trim();
+            String targetCurrency = parts[2].trim();
+            String dateString = parts[3];
+
+            double amount = parseAmount(amountAsString, line, lineNumber);
+            if (amount < 0) {
+                return false;
+            }
+
+            Optional<LocalDate> parsedDateOptional = parseDate(dateString, line, lineNumber);
+            LocalDate date;
+            if (parsedDateOptional.isPresent()) {
+                date = parsedDateOptional.get();
+            } else {
+                return false;
+            }
+
+            String formattedDate = date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));
+
+            sourceCurrency = validateSourceCurrency(line, lineNumber, sourceCurrency, formattedDate);
+            if (sourceCurrency == null) {
+                return false;
+            }
+
+            targetCurrency = validateTargetCurrency(line, lineNumber, targetCurrency, formattedDate);
+            if (targetCurrency == null) {
+                return false;
+            }
+
+            double exchangeRate = calculateExchangeRate(sourceCurrency, amount, targetCurrency, formattedDate, line);
+            // "_" + line  <- added, I have to check why its showing me result in different way
+            String uniqueKey = sourceCurrency + "_" + parts[1] + "_" + targetCurrency + "_" + dateString;
+
+            exchangeRatesFromFile.put(uniqueKey, exchangeRate);
+            return true;
+        } else {
+            recordError("Invalid line format in the file at line " + lineNumber, "Invalid line format in the file " + lineNumber + ": " + line);
+            return false;
+        }
+    }
+
+    private double parseAmount(String amountString, String line, int lineNumber) {
+        try {
+            return Double.parseDouble(amountString.trim());
+        } catch (NumberFormatException e) {
+            String errorKey = "Invalid amount in the file at line " + lineNumber;
+            recordError(errorKey, "Invalid amount in the file at line " + lineNumber + ": " + line);
+            return -1;
+        }
+    }
+
+    private Optional<LocalDate> parseDate(String dateString, String line, int lineNumber) {
+        try {
+            return Optional.of(LocalDate.parse(dateString, DateTimeFormatter.ofPattern("dd-MM-yy")));
+        } catch (DateTimeParseException e) {
+            recordError("Invalid date format in the file at line " + lineNumber, "Invalid date format in the file at line " + lineNumber + ": " + line);
+            return Optional.empty();
```

Comment:
Why is there an optional here, but every other method returns null?

Reply:
True, changed